clifford/__init__.py

```diff
@@ -193,53 +167,194 @@ def _get_mult_function(k_list, l_list, m_list, mult_table_vals, n_dims):
         The indices of the output
     m_list : array_like (N,) of intp
         The indices of the right operand
-    mult_table_vals : array_like (N,) of numbers.Real
+    val_list : array_like (N,) of numbers.Real
         The scalar multiplier.
-    n_dims: int
+    dims: int
         Length of the input and output vectors
-
-    Returns
-    -------
-    func : function (array_like (n_dims,), array_like (n_dims,)) -> array_like (n_dims,)
-        A function that computes the appropriate multiplication
     """
-    @numba.njit
-    def mv_mult(value, other_value):
-        output = np.zeros(n_dims)
-        for ind, k in enumerate(k_list):
-            m = m_list[ind]
-            l = l_list[ind]
-            output[l] += value[k] * mult_table_vals[ind] * other_value[m]
+    def __init__(self, k_list, l_list, m_list, val_list, dims):
+        self._k_list = k_list
+        self._l_list = l_list
+        self._m_list = m_list
+        self._val_list = val_list
+        self._dims = dims
+
+    # fake being an array, for compatibility with older code
+    @property
+    def ndim(self):
+        return 3
+
+    @property
+    def shape(self):
+        return (self._dims,)*3
+
+    @property
+    def dtype(self):
+        return self._val_list.dtype
+
+    def nonzero(self):
+        val_nonzero = self._val_list.nonzero()
+        return (
+            self._k_list[val_nonzero],
+            self._l_list[val_nonzero],
+            self._m_list[val_nonzero],
+        )
+
+    def __getitem__(self, item):
+        k, l, m = item
+        # add an axis to use for matching
+        k = np.asarray(k)[...,None]
+        l = np.asarray(l)[...,None]
+        m = np.asarray(m)[...,None]
+
+        # reshape the array to handle broadcasting
+        index_shape = np.broadcast(k, l, m, self._val_list).shape
+        broadcast_vals = np.broadcast_to(self._val_list, index_shape)
+
+        matching = (k == self._k_list) & (l == self._l_list) & (m == self._m_list)
+        try:
+            # numpy 1.17
+            return np.sum(broadcast_vals, where=matching, axis=-1)
+        except TypeError:
+            return np.sum(broadcast_vals*matching, axis=-1)
```

Comment:
gh-93 required `gmt[gmt.nonzero()]` to work - this makes that and variants behave correctly.